File: apps/envoygo/test/coding/coding_composer_test.dart
import 'package:envoygo/coding/coding_composer.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() async {
    SharedPreferences.setMockInitialValues({});
  });

  group('codingComposerCapabilities', () {
    test('publishes EH modes + permissions', () {
      final eh = codingComposerCapabilities('envoy-harness');
      expect(eh.agentModes.map((m) => m.id).toList(),
          ['default', 'plan', 'review']);
      expect(eh.canSetMode, isTrue);
      expect(eh.permissions, isTrue);
      expect(eh.workingMode, isFalse);
    });

    test('disables Full access for Codex sidecar', () {
      final codex = codingComposerCapabilities('codex');
      expect(codex.permissionFullDisabledReason, isNotNull);
      expect(codex.permissionAskDisabledReason, isNull);
      expect(codex.agentModes.any((m) => m.id == 'agent'), isTrue);
    });

    test('disables Ask for catalog ACP until a Mesh dock ships', () {
      final gemini = codingComposerCapabilities('gemini');
      expect(gemini.permissionAskDisabledReason, isNotNull);
      expect(gemini.permissionFullDisabledReason, isNull);
    });

    test('gates fast/plan/thinking for Codex and Claude Code', () {
      final codex = codingComposerCapabilities('codex');
      expect(codex.fast, isTrue);
      expect(codex.planSlash, isTrue);
      expect(codex.thinking, isTrue);

      final pi = codingComposerCapabilities('pi');
      expect(pi.fast, isFalse);
      expect(pi.planSlash, isFalse);
      expect(pi.permissions, isTrue);
    });
  });

  group('shapeCodingComposerPrompt', () {
    test('uses Review prefix for EH agent mode', () {
      final caps = codingComposerCapabilities('envoy-harness');
      final out = shapeCodingComposerPrompt(
        'look over auth',
        const CodingComposerPrefs(agentModeId: 'review'),
        caps,
      );
      expect(out, contains('[Mode: Review]'));
    });

    test('uses /plan when planSlash is available', () {
      final out = shapeCodingComposerPrompt(
        'refactor auth',
        const CodingComposerPrefs(mode: 'plan', agentModeId: 'plan'),
        codingComposerCapabilities('codex'),
      );
      expect(out, '/plan refactor auth');
    });

    test('leaves Code mode body unchanged', () {
      final caps = codingComposerCapabilities('gemini');
      expect(
        shapeCodingComposerPrompt(
          'fix it',
          const CodingComposerPrefs(mode: 'code'),
          caps,
        ),
        'fix it',
      );
    });
  });

  group('fastToggleSlash', () {
    test('emits on/off', () {
      expect(fastToggleSlash(true), '/fast on');
      expect(fastToggleSlash(false), '/fast off');
    });
  });

  group('codingComposerPrefs storage', () {
    test('persists mode, permissions, and agentModeId', () async {
      final key = codingComposerSessionKey(kind: 'ext', id: 'ext:1');
      await saveCodingComposerPrefs(
        key,
        const CodingComposerPrefs(
          mode: 'plan',
          fast: true,
          permissionPolicy: 'always-confirm',
          agentModeId: 'agent',
        ),
      );
      final loaded = await loadCodingComposerPrefs(key);
      expect(loaded.mode, 'plan');
      expect(loaded.fast, isTrue);
      expect(loaded.permissionPolicy, 'always-confirm');
      expect(loaded.agentModeId, 'agent');
    });
  });
}

File: apps/social/test/lib/coding-composer.test.ts
/**
 * @vitest-environment jsdom
 */
import { beforeEach, describe, expect, it } from "vitest";
import {
  codingComposerCapabilities,
  workingModeFromAgentMode,
} from "../../src/lib/coding-composer-capabilities.js";
import {
  shapeCodingComposerPrompt,
  fastToggleSlash,
} from "../../src/lib/coding-composer-prompt.js";
import {
  codingComposerSessionKey,
  loadCodingComposerPrefs,
  saveCodingComposerPrefs,
} from "../../src/lib/coding-composer-state.js";

describe("codingComposerCapabilities", () => {
  it("publishes EH modes + permissions", () => {
    const eh = codingComposerCapabilities("envoy-harness");
    expect(eh.agentModes.map((m) => m.id)).toEqual([
      "default",
      "plan",
      "review",
    ]);
    expect(eh.canSetMode).toBe(true);
    expect(eh.permissions).toBe(true);
    expect(eh.workingMode).toBe(false);
  });

  it("enables Pi permissions without native modes", () => {
    const pi = codingComposerCapabilities("pi");
    expect(pi.permissions).toBe(true);
    expect(pi.agentModes).toEqual([]);
    expect(pi.workingMode).toBe(true);
  });

  it("shows no Permissions control for a sidecar the coding run path cannot gate", () => {
    const codex = codingComposerCapabilities("codex");
    // The control is *absent*, not drawn with two values that change nothing: a sidecar backend
    // never reads `permissionPolicy`, so a select here would be the lie this table prevents.
    expect(codex.permissions).toBe(false);
    expect(codex.permissionDisabledReason).toMatch(/would change nothing/);
    expect(codex.permissionFullDisabledReason).toBeUndefined();
    expect(codex.agentModes.some((m) => m.id === "agent")).toBe(true);
    expect(codex.canSetMode).toBe(false);
  });

  it("gives a catalog ACP agent the full Permissions control, Ask included", () => {
    const gemini = codingComposerCapabilities("gemini");
    // The Mesh dock ships now (`coding:permission` → `codingRespondToPermission`), so "Ask every
    // time" is answerable instead of cancelling the tool — nothing here may disable it again
    // without also removing the dock.
    expect(gemini.permissions).toBe(true);
    expect(gemini.permissionAskDisabledReason).toBeUndefined();
    expect(gemini.permissionDisabledReason).toBeUndefined();
    expect(gemini.permissionFullDisabledReason).toBeUndefined();
  });

  it("gates fast/plan/thinking for Codex and Claude Code", () => {
    const codex = codingComposerCapabilities("codex");
    expect(codex.fast).toBe(true);
    expect(codex.planSlash).toBe(true);
    expect(codex.thinking).toBe(true);

    const pi = codingComposerCapabilities("pi");
    expect(pi.fast).toBe(false);
    expect(pi.planSlash).toBe(false);
  });
});

describe("shapeCodingComposerPrompt", () => {
  it("uses agent mode Review prefix for EH", () => {
    const caps = codingComposerCapabilities("envoy-harness");
    const out = shapeCodingComposerPrompt(
      "look over auth",
      {
        mode: "code",
        fast: false,
        thinking: "off",
        agentModeId: "review",
      },
      caps,
    );
    expect(out).toContain("[Mode: Review]");
    expect(workingModeFromAgentMode(caps, "plan")).toBe("plan");
  });

  it("uses /plan when planSlash is available", () => {
    const out = shapeCodingComposerPrompt(
      "refactor auth",
      { mode: "plan", fast: false, thinking: "off" },
      codingComposerCapabilities("codex"),
    );
    expect(out).toBe("/plan refactor auth");
  });

  it("leaves Code mode body unchanged for catalog without agentModes", () => {
    const caps = codingComposerCapabilities("gemini");
    expect(
      shapeCodingComposerPrompt(
        "fix it",
        { mode: "code", fast: false, thinking: "off" },
        caps,
      ),
    ).toBe("fix it");
  });
});

describe("fastToggleSlash", () => {
  it("emits on/off", () => {
    expect(fastToggleSlash(true)).toBe("/fast on");
    expect(fastToggleSlash(false)).toBe("/fast off");
  });
});

describe("codingComposerPrefs storage", () => {
  beforeEach(() => {
    localStorage.clear();
  });

  it("persists mode, permissions, and agentModeId", () => {
    const key = codingComposerSessionKey({ kind: "ext", id: "ext:1" });
    saveCodingComposerPrefs(key, {
      mode: "plan",
      fast: true,
      permissionPolicy: "always-confirm",
      agentModeId: "agent",
    });
    const loaded = loadCodingComposerPrefs(key);
    expect(loaded.mode).toBe("plan");
    expect(loaded.fast).toBe(true);
    expect(loaded.permissionPolicy).toBe("always-confirm");
    expect(loaded.agentModeId).toBe("agent");
  });
});
